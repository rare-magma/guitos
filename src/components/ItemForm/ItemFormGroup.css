.btn-delete {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
  color: var(--red);
}

.btn-delete:focus-visible,
.btn-delete:hover {
  background-color: var(--red);
  border: 1px solid var(--red);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}
