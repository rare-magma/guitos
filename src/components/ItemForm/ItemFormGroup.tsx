import { produce } from "immer";
import { RefObject, useRef, useState } from "react";
import {
  Button,
  Form,
  InputGroup,
  OverlayTrigger,
  Popover,
  Tooltip,
} from "react-bootstrap";
import CurrencyInput from "react-currency-input-field";
import { BsXLg } from "react-icons/bs";
import { useBudget } from "../../context/BudgetContext";
import { useConfig } from "../../context/ConfigContext";
import { useDB } from "../../hooks/useDB";
import {
  calc,
  calcAvailable,
  calcSaved,
  calcTotal,
  calcWithGoal,
  parseLocaleNumber,
  roundBig,
} from "../../utils";
import {
  CalculateButton,
  ItemOperation,
} from "../CalculateButton/CalculateButton";
import { Expense } from "../TableCard/Expense";
import { Income } from "../TableCard/Income";
import { ItemForm } from "./ItemForm";
import "./ItemFormGroup.css";

interface ItemFormProps {
  itemForm: ItemForm;
  costPercentage: number;
  label: string;
  inputRef: RefObject<HTMLInputElement>;
}

export function ItemFormGroup({
  itemForm,
  costPercentage,
  inputRef,
  label,
}: ItemFormProps) {
  const [needsRerender, setNeedsRerender] = useState(false);
  const deleteButtonRef = useRef<HTMLButtonElement>(null);
  const valueRef = useRef<HTMLInputElement>(null);
  const { budget, setBudget } = useBudget();
  const { deleteCalcHist, saveCalcHist } = useDB();
  const { intlConfig } = useConfig();
  const isExpense = label === "Expenses";
  const table = isExpense ? budget?.expenses : budget?.incomes;

  function handleCalcHist(
    operation: ItemOperation,
    changeValue: number | undefined,
  ) {
    if (!budget) return;
    const newItemForm = isExpense
      ? budget.expenses.items.find((item) => item.id === itemForm.id)
      : budget.incomes.items.find((item) => item.id === itemForm.id);
    if (!newItemForm) return;
    const calcHistID = `${budget.id}-${label}-${newItemForm.id}`;
    saveCalcHist(calcHistID, {
      id: calcHistID,
      itemForm: newItemForm,
      changeValue,
      operation,
    }).catch((e: unknown) => {
      throw e;
    });
  }

  function handleChange(
    operation: ItemOperation,
    value?: string,
    event?: React.ChangeEvent<HTMLInputElement>,
    changeValue?: number,
  ) {
    if (!budget) return;
    if (!itemForm) return;

    let saveInHistory = false;
    const newState = produce((draft) => {
      const newItemForm = isExpense
        ? draft.expenses.items.find((item) => item.id === itemForm.id)
        : draft.incomes.items.find((item) => item.id === itemForm.id);
      if (!newItemForm) return;

      switch (operation) {
        case "name":
          if (event) newItemForm.name = event.target.value;
          break;
        case "value":
          if (value) {
            newItemForm.value = parseLocaleNumber(value, intlConfig?.locale);
          }
          break;
        default:
          if (changeValue) {
            newItemForm.value = calc(itemForm.value, changeValue, operation);
            saveInHistory = true;
          }
          setNeedsRerender(!needsRerender);
          handleCalcHist(operation, changeValue);
          break;
      }

      isExpense
        ? (draft.expenses.total = roundBig(calcTotal(draft.expenses.items), 2))
        : (draft.incomes.total = roundBig(calcTotal(draft.incomes.items), 2));
      draft.stats.available = roundBig(calcAvailable(draft), 2);
      draft.stats.withGoal = calcWithGoal(draft);
      draft.stats.saved = calcSaved(draft);
    }, budget);

    setBudget(newState(), saveInHistory);
  }

  function handleRemove(toBeDeleted: ItemForm) {
    if (!table?.items) return;
    if (!budget) return;

    const newTable = isExpense ? ({} as Expense) : ({} as Income);
    const newState = produce((draft) => {
      isExpense ? (draft.expenses = newTable) : (draft.incomes = newTable);
      newTable.items = table.items.filter((item) => item.id !== toBeDeleted.id);
      newTable.total = roundBig(calcTotal(newTable.items), 2);
      draft.stats.available = roundBig(calcAvailable(draft), 2);
      draft.stats.withGoal = calcWithGoal(draft);
      draft.stats.saved = calcSaved(draft);
    }, budget);
    setBudget(newState(), true);

    const calcHistID = `${budget.id}-${label}-${toBeDeleted.id}`;
    deleteCalcHist(calcHistID).catch((e: unknown) => {
      throw e;
    });
  }

  return (
    <InputGroup
      size="sm"
      className="mb-1"
      key={`${itemForm.id}-${label}-group`}
    >
      <OverlayTrigger
        delay={250}
        placement="top"
        overlay={
          costPercentage > 0 ? (
            <Tooltip
              id={`tooltip-value-${label}-${itemForm.id}`}
              style={{ position: "fixed" }}
            >
              {costPercentage}% of revenue
            </Tooltip>
          ) : (
            <></>
          )
        }
      >
        <Form.Control
          id={`${label}-${itemForm.id}-name`}
          aria-label={`item ${itemForm.id} name`}
          key={`${itemForm.id}-${label}-name`}
          className="w-25"
          ref={inputRef}
          defaultValue={itemForm.name}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            handleChange("name", undefined, e)
          }
          type="text"
          maxLength={255}
        />
      </OverlayTrigger>
      <OverlayTrigger
        delay={250}
        placement="top"
        overlay={
          costPercentage > 0 ? (
            <Tooltip
              id={`tooltip-costPercentage-${label}-${itemForm.id}`}
              style={{ position: "fixed" }}
            >
              {costPercentage}% of revenue
            </Tooltip>
          ) : (
            <></>
          )
        }
      >
        <CurrencyInput
          id={`${label}-${itemForm.id}-value`}
          key={`${itemForm.id}-${label}-value-${needsRerender}`}
          className="text-end form-control straight-corners fixed-width-font"
          aria-label={`item ${itemForm.id} value`}
          name="item-value"
          intlConfig={intlConfig}
          defaultValue={itemForm.value}
          allowNegativeValue={false}
          maxLength={14}
          ref={valueRef}
          onFocus={() => valueRef.current?.setSelectionRange(0, 25)}
          onValueChange={(value) => handleChange("value", value)}
        />
      </OverlayTrigger>
      <CalculateButton
        itemForm={itemForm}
        label={label}
        onCalculate={(changeValue, operation) =>
          handleChange(operation, "", undefined, changeValue)
        }
      />
      <OverlayTrigger
        trigger="click"
        placement="top"
        rootClose={true}
        overlay={
          <Popover id={`popover-delete-button`}>
            <Popover.Body>
              <OverlayTrigger
                delay={250}
                placement="top"
                overlay={
                  <Tooltip
                    id={`tooltip-delete-itemformgroup`}
                    style={{ position: "fixed" }}
                  >
                    delete item
                  </Tooltip>
                }
              >
                <Button
                  id={`item-${itemForm.id}-delete-confirmation-button`}
                  aria-label={`confirm item ${itemForm.id} deletion`}
                  key={`${itemForm.id}-${label}-delete-confirmation-button`}
                  variant="delete"
                  type="button"
                  size="sm"
                  ref={deleteButtonRef}
                  onClick={() => handleRemove(itemForm)}
                >
                  <BsXLg aria-hidden />
                </Button>
              </OverlayTrigger>
            </Popover.Body>
          </Popover>
        }
      >
        <Button
          id={`delete-${label}-${itemForm.id}-button`}
          aria-label={`delete item ${itemForm.id}`}
          key={`${itemForm.id}-${label}-delete-button`}
          variant="delete"
          type="button"
          onClick={() => {
            setTimeout(() => {
              deleteButtonRef.current?.focus();
            }, 0);
          }}
        >
          <BsXLg aria-hidden />
        </Button>
      </OverlayTrigger>
    </InputGroup>
  );
}
