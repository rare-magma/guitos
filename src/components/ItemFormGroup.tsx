import { useState } from "react";
import { Button, Form, InputGroup } from "react-bootstrap";
import { ItemForm } from "./ItemForm";

interface ItemFormProps {
  itemForm: ItemForm;
  onRemove: (itemForm: ItemForm) => void;
}

function ItemFormGroup({ itemForm: initialItemForm, onRemove }: ItemFormProps) {
  const [itemForm, setItemForm] = useState(initialItemForm);

  const handleRemove = (item: ItemForm) => {
    onRemove(item);
  };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const handleChange = (event: any) => {
    const { type, name, value } = event.target;
    let updatedValue = value;

    if (type === "number") {
      updatedValue = Number(updatedValue);
    }
    const change = {
      [name]: updatedValue,
    };

    let updatedItemForm: ItemForm;

    setItemForm((i) => {
      updatedItemForm = new ItemForm({ ...i, ...change });
      return updatedItemForm;
    });
  };

  return (
    <Form>
      <InputGroup size="sm" className="mb-1" key={itemForm.id + "-group"}>
        <Form.Control
          aria-label={"newname"}
          key={itemForm.id + "-key"}
          defaultValue={itemForm.name}
          onChange={handleChange}
          type="text"
          maxLength={255}
        />
        <Form.Control
          aria-label={"newvalue"}
          key={itemForm.id + "-value"}
          className="text-right"
          defaultValue={itemForm.value}
          onChange={handleChange}
          type="number"
          maxLength={11}
        />
        <InputGroup.Text>â‚¬</InputGroup.Text>
        <Button
          key={itemForm.id + "button"}
          variant="text"
          type="button"
          onClick={() => {
            handleRemove(itemForm);
          }}
        >
          -
        </Button>
      </InputGroup>
    </Form>
  );
}

export default ItemFormGroup;
