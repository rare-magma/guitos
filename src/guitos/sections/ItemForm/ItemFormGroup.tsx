import { immerable, produce } from "immer";
import type React from "react";
import { type RefObject, useRef, useState } from "react";
import {
  Button,
  Form,
  InputGroup,
  OverlayTrigger,
  Popover,
  Tooltip,
} from "react-bootstrap";
import CurrencyInput from "react-currency-input-field";
import { BsXLg } from "react-icons/bs";
import { calc, parseLocaleNumber, roundBig } from "../../../utils";
import { useBudget } from "../../context/BudgetContext";
import type { Expenses } from "../../domain/expenses";
import type { Incomes } from "../../domain/incomes";
import { useDB } from "../../hooks/useDB";
import { CalculateButton } from "../CalculateButton/CalculateButton";
import "./ItemFormGroup.css";
import { Budget } from "../../domain/budget";
import type { BudgetItem } from "../../domain/budgetItem";
import type { ItemOperation } from "../../domain/calculationHistoryItem";
import { UserOptions } from "../../domain/userOptions";

interface ItemFormProps {
  itemForm: BudgetItem;
  costPercentage: number;
  label: string;
  inputRef: RefObject<HTMLInputElement>;
  userOptions: UserOptions;
}

export function ItemFormGroup({
  itemForm,
  costPercentage,
  inputRef,
  label,
  userOptions,
}: ItemFormProps) {
  const [needsRerender, setNeedsRerender] = useState(false);
  const deleteButtonRef = useRef<HTMLButtonElement>(null);
  const valueRef = useRef<HTMLInputElement>(null);
  const { budget, setBudget } = useBudget();
  const { deleteCalcHist, saveCalcHist } = useDB();
  const isExpense = label === "Expenses";
  const table = isExpense ? budget?.expenses : budget?.incomes;

  function handleCalcHist(operation: ItemOperation, changeValue: number) {
    if (!budget) return;
    const newItemForm = isExpense
      ? budget.expenses.items.find(
          (item: BudgetItem) => item.id === itemForm.id,
        )
      : budget.incomes.items.find(
          (item: BudgetItem) => item.id === itemForm.id,
        );
    if (!newItemForm) return;
    const calcHistID = `${budget.id}-${label}-${newItemForm.id}`;
    saveCalcHist(calcHistID, {
      [immerable]: true,
      id: calcHistID,
      itemForm: newItemForm,
      changeValue,
      operation,
    }).catch((e: unknown) => {
      throw e;
    });
  }

  function handleChange(
    operation: ItemOperation,
    value?: string,
    event?: React.ChangeEvent<HTMLInputElement>,
    changeValue?: number,
  ) {
    if (!budget) return;
    if (!itemForm) return;

    let saveInHistory = false;
    // biome-ignore lint/complexity/noExcessiveCognitiveComplexity: <explanation>
    const newState = produce((draft) => {
      const newItemForm = isExpense
        ? draft.expenses.items.find(
            (item: BudgetItem) => item.id === itemForm.id,
          )
        : draft.incomes.items.find(
            (item: BudgetItem) => item.id === itemForm.id,
          );
      if (!newItemForm) return;

      switch (operation) {
        case "name":
          if (event) newItemForm.name = event.target.value;
          break;
        case "value":
          if (value) {
            newItemForm.value = parseLocaleNumber(value, userOptions.locale);
          }
          break;
        default:
          if (changeValue) {
            newItemForm.value = calc(itemForm.value, changeValue, operation);
            saveInHistory = true;
            setNeedsRerender(!needsRerender);
            handleCalcHist(operation, changeValue);
          }
          break;
      }

      if (isExpense) {
        draft.expenses.total = roundBig(
          Budget.itemsTotal(draft.expenses.items),
          2,
        );
      } else {
        draft.incomes.total = roundBig(
          Budget.itemsTotal(draft.incomes.items),
          2,
        );
      }
      draft.stats.available = roundBig(Budget.available(draft as Budget), 2);
      draft.stats.withGoal = Budget.availableWithGoal(draft as Budget);
      draft.stats.saved = Budget.saved(draft as Budget);
    }, budget);

    setBudget(newState(), saveInHistory);
  }

  function handleRemove(toBeDeleted: BudgetItem) {
    if (!table?.items) return;
    if (!budget) return;

    const newTable = isExpense ? ({} as Expenses) : ({} as Incomes);
    const newState = produce((draft) => {
      if (isExpense) {
        draft.expenses = newTable;
      } else {
        draft.incomes = newTable;
      }
      newTable.items = table.items.filter(
        (item: BudgetItem) => item.id !== toBeDeleted.id,
      );
      newTable.total = roundBig(Budget.itemsTotal(newTable.items), 2);
      draft.stats.available = roundBig(Budget.available(draft as Budget), 2);
      draft.stats.withGoal = Budget.availableWithGoal(draft as Budget);
      draft.stats.saved = Budget.saved(draft as Budget);
    }, budget);
    setBudget(newState(), true);

    const calcHistID = `${budget.id}-${label}-${toBeDeleted.id}`;
    deleteCalcHist(calcHistID).catch((e: unknown) => {
      throw e;
    });
  }

  return (
    <InputGroup
      size="sm"
      className="mb-1"
      key={`${itemForm.id}-${label}-group`}
    >
      <OverlayTrigger
        delay={250}
        placement="top"
        overlay={
          costPercentage > 0 ? (
            <Tooltip
              id={`tooltip-value-${label}-${itemForm.id}`}
              style={{ position: "fixed" }}
            >
              {costPercentage}% of revenue
            </Tooltip>
          ) : (
            <></>
          )
        }
      >
        <Form.Control
          id={`${label}-${itemForm.id}-name`}
          aria-label={`item ${itemForm.id} name`}
          key={`${itemForm.id}-${label}-name`}
          className="w-25"
          ref={inputRef}
          defaultValue={itemForm.name}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            handleChange("name", undefined, e)
          }
          type="text"
          maxLength={255}
        />
      </OverlayTrigger>
      <OverlayTrigger
        delay={250}
        placement="top"
        overlay={
          costPercentage > 0 ? (
            <Tooltip
              id={`tooltip-costPercentage-${label}-${itemForm.id}`}
              style={{ position: "fixed" }}
            >
              {costPercentage}% of revenue
            </Tooltip>
          ) : (
            <></>
          )
        }
      >
        <CurrencyInput
          id={`${label}-${itemForm.id}-value`}
          key={`${itemForm.id}-${label}-value-${needsRerender}`}
          className="text-end form-control straight-corners fixed-width-font"
          aria-label={`item ${itemForm.id} value`}
          name="item-value"
          intlConfig={UserOptions.toIntlConfig(userOptions)}
          defaultValue={itemForm.value}
          allowNegativeValue={false}
          maxLength={14}
          ref={valueRef}
          onFocus={() => valueRef.current?.setSelectionRange(0, 25)}
          onValueChange={(value) => handleChange("value", value)}
        />
      </OverlayTrigger>
      <CalculateButton
        itemForm={itemForm}
        label={label}
        onCalculate={(changeValue, operation) =>
          handleChange(operation, "", undefined, changeValue)
        }
      />
      <OverlayTrigger
        trigger="click"
        placement="top"
        rootClose={true}
        overlay={
          <Popover id={"popover-delete-button"}>
            <Popover.Body>
              <OverlayTrigger
                delay={250}
                placement="top"
                overlay={
                  <Tooltip
                    id={"tooltip-delete-itemformgroup"}
                    style={{ position: "fixed" }}
                  >
                    delete item
                  </Tooltip>
                }
              >
                <Button
                  id={`item-${itemForm.id}-delete-confirmation-button`}
                  aria-label={`confirm item ${itemForm.id} deletion`}
                  key={`${itemForm.id}-${label}-delete-confirmation-button`}
                  variant="delete"
                  type="button"
                  size="sm"
                  ref={deleteButtonRef}
                  onClick={() => handleRemove(itemForm)}
                >
                  <BsXLg aria-hidden={true} />
                </Button>
              </OverlayTrigger>
            </Popover.Body>
          </Popover>
        }
      >
        <Button
          id={`delete-${label}-${itemForm.id}-button`}
          aria-label={`delete item ${itemForm.id}`}
          key={`${itemForm.id}-${label}-delete-button`}
          variant="delete"
          type="button"
          onClick={() => {
            setTimeout(() => {
              deleteButtonRef.current?.focus();
            }, 0);
          }}
        >
          <BsXLg aria-hidden={true} />
        </Button>
      </OverlayTrigger>
    </InputGroup>
  );
}
