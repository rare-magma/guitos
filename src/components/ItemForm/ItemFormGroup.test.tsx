import { cleanup, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import {
  configContextSpy,
  itemForm1,
  setBudgetMock,
  testBudget,
  testSpanishConfigContext,
} from "../../setupTests";
import { ItemFormGroup } from "./ItemFormGroup";

describe("ItemFormGroup", () => {
  const ref = React.createRef<HTMLInputElement>();
  const comp = (
    <ItemFormGroup
      itemForm={itemForm1}
      label="Expenses"
      inputRef={ref}
      costPercentage={1}
    />
  );

  it("matches snapshot", () => {
    render(comp);
    expect(comp).toMatchSnapshot();
  });

  it("renders initial state", async () => {
    render(comp);
    expect(await screen.findByDisplayValue("name1")).toBeInTheDocument();
    expect(await screen.findByDisplayValue("$10")).toBeInTheDocument();
  });

  it("reacts to user changing input", async () => {
    render(comp);
    setBudgetMock.mockClear();
    await waitFor(async () => {
      await userEvent.type(screen.getByDisplayValue("name1"), "change name");
    });

    expect(screen.getByDisplayValue("name1change name")).toBeInTheDocument();
    expect(setBudgetMock).toHaveBeenCalledWith(
      {
        ...testBudget,
        expenses: {
          items: [{ id: 1, name: "name1change name", value: 10 }],
          total: 10,
        },
      },
      false,
    );

    setBudgetMock.mockClear();

    await waitFor(async () => {
      await userEvent.type(screen.getByDisplayValue("$10"), "123");
    });

    expect(screen.getByDisplayValue("$123")).toBeInTheDocument();
    expect(setBudgetMock).toHaveBeenCalledWith(
      {
        ...testBudget,
        expenses: {
          items: [{ id: 1, name: "expense1", value: 123 }],
          total: 123,
        },
        stats: {
          ...testBudget.stats,
          available: -23,
          withGoal: -33,
        },
      },
      false,
    );
  });

  it("removes item when user clicks delete confirmation button", async () => {
    render(comp);
    setBudgetMock.mockClear();
    await waitFor(async () => {
      await userEvent.click(
        screen.getByRole("button", { name: "delete item 1" }),
      );
      await userEvent.click(
        screen.getByRole("button", { name: "confirm item 1 deletion" }),
      );
    });

    expect(setBudgetMock).toHaveBeenCalledWith(
      {
        ...testBudget,
        expenses: { items: [], total: 0 },
        stats: {
          ...testBudget.stats,
          available: 100,
          withGoal: 90,
        },
      },
      true,
    );
  });

  it("shows tooltip when user hovers over", async () => {
    render(comp);
    await waitFor(async () => {
      await userEvent.hover(screen.getByDisplayValue("$10"));
    });

    expect(await screen.findByText("1% of revenue")).toBeInTheDocument();
  });

  it("opens popover when clicking the button", async () => {
    render(comp);
    await waitFor(async () => {
      await userEvent.click(
        screen.getByRole("button", {
          name: "select operation type to item value",
        }),
      );
    });
    expect(
      screen.getByLabelText("select type of operation on item value"),
    ).toBeInTheDocument();
  });

  it("transforms decimal separator based on locale", async () => {
    cleanup();

    configContextSpy.mockReturnValue(testSpanishConfigContext);

    render(
      <ItemFormGroup
        itemForm={itemForm1}
        label="Expenses"
        inputRef={ref}
        costPercentage={1}
      />,
    );

    await waitFor(async () => {
      await userEvent.clear(screen.getByDisplayValue("10 €"));
      await userEvent.type(screen.getByLabelText("item 1 value"), ",12");
    });

    expect(screen.getByDisplayValue("0,12 €")).toBeInTheDocument();
  });
});
