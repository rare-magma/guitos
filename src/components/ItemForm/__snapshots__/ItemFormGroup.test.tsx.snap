// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ItemFormGroup > matches snapshot 1`] = `
<ItemFormGroup
  costPercentage={1}
  inputRef={
    {
      "current": <input
        aria-label="item 1 name"
        class="w-25 form-control"
        id="Expenses-1-name"
        maxlength="255"
        type="text"
        value="name1"
      />,
    }
  }
  itemForm={
    ItemForm {
      "id": 1,
      "name": "name1",
      "value": 10,
    }
  }
  label="Expenses"
  onChange={[MockFunction spy]}
  onRemove={[MockFunction spy]}
/>
`;
